File: Util/JavaCryptanalysis/src/m209/SimulatedAnnealingPins.java
package m209;

import common.SimulatedAnnealing;

public class SimulatedAnnealingPins {

    private static int step(int task, int[] roundLayers, int layers, Key key, EvalType evalType,
                            boolean[][] bestSAPins, int bestSAScore,
                            double temperature) {


        int currLocalScore = 0;

        final int MAX_COUNT = key.pins.maxCount();
        final int MIN_COUNT = key.pins.minCount();

        int newScore;
        int count;

        boolean changeAccepted = false;

        for (int w = 1; w <= Key.WHEELS; w++) {
            for (int p = 0; p < Key.WHEELS_SIZE[w]; p++) {
                key.pins.toggle(w, p);
                count = key.pins.count();
                if (count < MIN_COUNT || count > MAX_COUNT || key.pins.longSeq(w, p)) {
                    key.pins.toggle(w, p);
                    continue;
                }
                key.updateDecryption(w, p);

                newScore = key.eval(evalType);
                if (SimulatedAnnealing.accept(newScore, currLocalScore, temperature)) {
                    currLocalScore = newScore;
                    changeAccepted = true;
                    if (newScore > bestSAScore) {
                        key.pins.get(bestSAPins);
                        bestSAScore = newScore;
                    }
                    ReportResult.reportResult(task, roundLayers, layers, key, currLocalScore, "SA P. x1");
                } else {
                    key.pins.toggle(w, p);
                    key.updateDecryption(w, p);
                }
            }
        }

        for (int w = 1; w <= Key.WHEELS; w++) {

            key.pins.inverse(w);
            count = key.pins.count();
            if (count < MIN_COUNT || count > MAX_COUNT) {
                key.pins.inverse(w);
                continue;
            }
            //key.updateDecryption();

            newScore = key.eval(evalType);
            if (SimulatedAnnealing.accept(newScore, currLocalScore, temperature)) {
                currLocalScore = newScore;
                changeAccepted = true;
                if (newScore > bestSAScore) {
                    key.pins.get(bestSAPins);
                    bestSAScore = newScore;
                }
                ReportResult.reportResult(task, roundLayers, layers, key, currLocalScore, "SA P. xw");

            } else {
                key.pins.inverse(w);
                //key.updateDecryption();
            }
        }

        if (changeAccepted) {
            return bestSAScore;
        }

        for (int w = 1; w <= Key.WHEELS; w++) {
            for (int p1 = 0; p1 < Key.WHEELS_SIZE[w]; p1++) {
                for (int p2 = p1 + 1; p2 < Key.WHEELS_SIZE[w]; p2++) {
                    if (key.pins.compare(w, p1, p2)) {
                        continue;
                    }
                    // Because the two places have a different value, we do not check the count.

                    key.pins.toggle(w, p1, p2);

                    if (key.pins.longSeq(w, p1, p2)) {
                        key.pins.toggle(w, p1, p2);
                        continue;
                    }
                    key.updateDecryption(w, p1, p2);

                    newScore = key.eval(evalType);

                    if (SimulatedAnnealing.accept(newScore, currLocalScore, temperature)) {
                        currLocalScore = newScore;
                        changeAccepted = true;
                        if (newScore > bestSAScore) {
                            key.pins.get(bestSAPins);
                            bestSAScore = newScore;
                        }
                        ReportResult.reportResult(task, roundLayers, layers, key, currLocalScore, "SA P. x2");
                    } else {
                        key.pins.toggle(w, p1, p2);
                        key.updateDecryption(w, p1, p2);
                    }

                }
            }
        }

        if (changeAccepted) {
            return bestSAScore;
        }
        int bestV = -1;
        int bestVscore = 0;
        for (int v = 0; v <= 63; v++) {
            key.pins.inverseWheelBitmap(v);
            int score = key.eval(evalType);
            if (score > bestVscore) {
                bestVscore = score;
                bestV = v;
            }
            key.pins.inverseWheelBitmap(v);
        }
        newScore = bestVscore;
        if (SimulatedAnnealing.accept(newScore, currLocalScore, temperature)) {
            currLocalScore = newScore;
            key.pins.inverseWheelBitmap(bestV);
            changeAccepted = true;
            if (newScore > bestSAScore) {
                key.pins.get(bestSAPins);
                bestSAScore = newScore;
            }
            ReportResult.reportResult(task, roundLayers, layers, key, currLocalScore, "SA P. xn");
        }

        return bestSAScore;

    }


    public static int SA(int task, int[] roundLayers, int layers, Key key, EvalType evalType, int cycles) {

        int bestSAScore = key.eval(evalType);
        boolean[][] bestSAPins = key.pins.createCopy();

        for (int cycle = 0; cycle < cycles; cycle++) {
            roundLayers[layers] = cycle;
            key.pins.randomize();

            roundLayers[layers + 1] = 0;
            for (double temperature = 1000.0; temperature >= 1.0; temperature /= 1.1, roundLayers[layers + 1]++) {
                bestSAScore = step(task, roundLayers, layers + 2, key, evalType, bestSAPins, bestSAScore, temperature);
            }
            long previous;
            do {
                previous = bestSAScore;
                bestSAScore = step(task, roundLayers, layers + 2, key, evalType, bestSAPins, bestSAScore, 0.0);
            } while (bestSAScore > previous);
        }

        key.pins.set(bestSAPins);
        return bestSAScore;

    }

}

File: LibSource/M209Analyzer/M209Analyzer/m209/SimulatedAnnealingPins.cs
/*
   Copyright CrypTool 2 Team josef.matwich@gmail.com

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/
using M209AnalyzerLib.Enums;

namespace M209AnalyzerLib.M209
{
    public class SimulatedAnnealingPins
    {
        private M209AttackManager _attackManager;
        private Key _key;
        private LocalState _localState;
        public SimulatedAnnealingPins(Key key, M209AttackManager attackManager, LocalState localState)
        {
            _key = key;
            _attackManager = attackManager;
            _localState = localState;
        }
        private double Step(EvalType evalType,
                                bool[][] bestSAPins, double bestSAScore,
                                double temperature)
        {
            double currLocalScore = 0;
            int MAX_COUNT = _key.Pins.MaxCount();
            int MIN_COUNT = _key.Pins.MinCount();

            double newScore;
            int count;

            bool changeAccepted = false;

            // T1
            for (int w = 1; w <= _key.WHEELS; w++)
            {
                for (int p = 0; p < _key.WHEELS_SIZE[w]; p++)
                {
                    _key.Pins.Toggle(w, p);
                    count = _key.Pins.CountActivePins();
                    if (count < MIN_COUNT || count > MAX_COUNT || _key.Pins.LongSeq(w, p))
                    {
                        _key.Pins.Toggle(w, p);
                        continue;
                    }
                    _key.UpdateDecryption(w, p);

                    newScore = _attackManager.Evaluate(evalType, _key.Decryption, _key.CribArray, _localState.TaskId);
                    if (_attackManager.SimulatedAnnealing.Accept(newScore, currLocalScore, temperature))
                    {
                        currLocalScore = newScore;
                        changeAccepted = true;
                        if (newScore > bestSAScore)
                        {
                            _key.Pins.Get(bestSAPins);
                            bestSAScore = newScore;
                        }
                        _attackManager.AddNewBestListEntry(currLocalScore, _key, _key.Decryption, _localState.TaskId);
                    }
                    else
                    {
                        _key.Pins.Toggle(w, p);
                        _key.UpdateDecryption(w, p);
                    }
                    if (_attackManager.ShouldStop)
                    {
                        return bestSAScore;
                    }
                }
            }

            // T4
            for (int w = 1; w <= _key.WHEELS; w++)
            {

                _key.Pins.Inverse(w);
                count = _key.Pins.CountActivePins();
                if (count < MIN_COUNT || count > MAX_COUNT)
                {
                    _key.Pins.Inverse(w);
                    continue;
                }
                _key.UpdateDecryptionIfInvalid();
                newScore = _attackManager.Evaluate(evalType, _key.Decryption, _key.CribArray, _localState.TaskId);
                if (_attackManager.SimulatedAnnealing.Accept(newScore, currLocalScore, temperature))
                {
                    currLocalScore = newScore;
                    changeAccepted = true;
                    if (newScore > bestSAScore)
                    {
                        _key.Pins.Get(bestSAPins);
                        bestSAScore = newScore;
                    }
                    _attackManager.AddNewBestListEntry(currLocalScore, _key, _key.Decryption, _localState.TaskId);

                }
                else
                {
                    _key.Pins.Inverse(w);
                }
                if (_attackManager.ShouldStop)
                {
                    return bestSAScore;
                }
            }

            if (changeAccepted | _attackManager.ShouldStop)
            {
                return bestSAScore;
            }

            // T2
            for (int w = 1; w <= _key.WHEELS; w++)
            {
                for (int p1 = 0; p1 < _key.WHEELS_SIZE[w]; p1++)
                {
                    for (int p2 = p1 + 1; p2 < _key.WHEELS_SIZE[w]; p2++)
                    {
                        if (_key.Pins.Compare(w, p1, p2))
                        {
                            continue;
                        }
                        // Because the two places have a different value, we do not check the count.

                        _key.Pins.Toggle(w, p1, p2);

                        if (_key.Pins.LongSeq(w, p1, p2))
                        {
                            _key.Pins.Toggle(w, p1, p2);
                            continue;
                        }
                        _key.UpdateDecryption(w, p1, p2);

                        newScore = _attackManager.Evaluate(evalType, _key.Decryption, _key.CribArray, _localState.TaskId);
                        if (_attackManager.SimulatedAnnealing.Accept(newScore, currLocalScore, temperature))
                        {
                            currLocalScore = newScore;
                            changeAccepted = true;
                            if (newScore > bestSAScore)
                            {
                                _key.Pins.Get(bestSAPins);
                                bestSAScore = newScore;
                            }
                            _attackManager.AddNewBestListEntry(currLocalScore, _key, _key.Decryption, _localState.TaskId);
                        }
                        else
                        {
                            _key.Pins.Toggle(w, p1, p2);
                            _key.UpdateDecryption(w, p1, p2);
                        }

                        if (_attackManager.ShouldStop)
                        {
                            return bestSAScore;
                        }
                    }
                }
            }

            if (changeAccepted | _attackManager.ShouldStop)
            {
                return bestSAScore;
            }

            // T?
            int bestV = -1;
            double bestVscore = 0;
            for (int v = 0; v <= 63; v++)
            {
                _key.Pins.InverseWheelBitmap(v);
                double score = _key.Eval(evalType);
                score = _attackManager.Evaluate(evalType, _key.Decryption, _key.CribArray, _localState.TaskId);
                if (score > bestVscore)
                {
                    bestVscore = score;
                    bestV = v;
                }
                _key.Pins.InverseWheelBitmap(v);
            }
            newScore = bestVscore;
            if (_attackManager.SimulatedAnnealing.Accept(newScore, currLocalScore, temperature))
            {
                currLocalScore = newScore;
                _key.Pins.InverseWheelBitmap(bestV);
                changeAccepted = true;
                if (newScore > bestSAScore)
                {
                    _key.Pins.Get(bestSAPins);
                    bestSAScore = newScore;
                }
                _attackManager.AddNewBestListEntry(currLocalScore, _key, _key.Decryption, _localState.TaskId);
            }

            return bestSAScore;

        }


        public double SA(EvalType evalType, int cycles)
        {
            _key.UpdateDecryptionIfInvalid();
            double bestSAScore = _attackManager.Evaluate(evalType, _key.Decryption, _key.CribArray, _localState.TaskId);
            bool[][] bestSAPins = _key.Pins.CreateCopy();

            for (int cycle = 0; cycle < cycles && !_attackManager.ShouldStop; cycle++)
            {
                _key.Pins.Randomize();

                for (double temperature = _attackManager.SAParameters.StartTemperature; temperature >= _attackManager.SAParameters.EndTemperature && !_attackManager.ShouldStop; temperature /= _attackManager.SAParameters.Decrement)
                {
                    bestSAScore = Step(evalType, bestSAPins, bestSAScore, temperature);
                }
                double previous;
                do
                {
                    previous = bestSAScore;
                    bestSAScore = Step(evalType, bestSAPins, bestSAScore, 0.0);
                } while (bestSAScore > previous && !_attackManager.ShouldStop);
            }

            _key.Pins.Set(bestSAPins);
            return bestSAScore;
        }
    }
}
